File: core/src/com/bombbird/terminalcontrol/utilities/GameLoader.java
package com.bombbird.terminalcontrol.utilities;

import com.badlogic.gdx.Gdx;
import com.bombbird.terminalcontrol.TerminalControl;
import com.bombbird.terminalcontrol.entities.Airport;
import com.bombbird.terminalcontrol.entities.Runway;
import com.bombbird.terminalcontrol.entities.aircrafts.Aircraft;
import com.bombbird.terminalcontrol.entities.aircrafts.Arrival;
import com.bombbird.terminalcontrol.entities.aircrafts.Departure;
import com.bombbird.terminalcontrol.entities.trafficmanager.ArrivalManager;
import org.json.JSONArray;
import org.json.JSONObject;

public class GameLoader {
    /** Loads save information from the save JSONObject */
    public static void loadSaveData(JSONObject save) {
        if (save == null) return;

        JSONArray airports = save.getJSONArray("airports");

        for (Airport airport: TerminalControl.radarScreen.airports.values()) {
            for (Runway runway: airport.getRunways().values()) {
                runway.getLabel().remove();
            }
        }
        loadAirportData(airports);
        TerminalControl.radarScreen.getMetar().updateMetar();
        loadAircraft(save.getJSONArray("aircrafts"));

        for (int i = 0; i < airports.length(); i++) {
            Airport airport = TerminalControl.radarScreen.airports.get(airports.getJSONObject(i).getString("icao"));
            airport.getTakeoffManager().updatePrevAcft(airports.getJSONObject(i).getJSONObject("takeoffManager"));
            airport.updateRunwayQueue(airports.getJSONObject(i));
        }

        TerminalControl.radarScreen.setArrivalManager(new ArrivalManager(save.getJSONObject("arrivalManager")));

        //GameSaver.saveGame();
    }

    /** Loads aircraft data from save */
    private static void loadAircraft(JSONArray aircrafts) {
        for (Aircraft aircraft: TerminalControl.radarScreen.aircrafts.values()) {
            aircraft.removeAircraft();
        }

        for (int i = 0; i < aircrafts.length(); i++) {
            if ("Arrival".equals(aircrafts.getJSONObject(i).getString("TYPE"))) {
                //Load arrival
                Arrival arrival = new Arrival(aircrafts.getJSONObject(i));
                TerminalControl.radarScreen.aircrafts.put(arrival.getCallsign(), arrival);
            } else if ("Departure".equals(aircrafts.getJSONObject(i).getString("TYPE"))) {
                //Load departure
                Departure departure = new Departure(aircrafts.getJSONObject(i));
                TerminalControl.radarScreen.aircrafts.put(departure.getCallsign(), departure);
            } else {
                //Unknown type TODO En-route aircraft in future
                Gdx.app.log("Aircraft load error", "Unknown aircraft type " + aircrafts.getJSONObject(i).getString("TYPE") + " in save file!");
            }
        }
    }

    /** Loads airport data from save */
    private static void loadAirportData(JSONArray airports) {
        for (int i = 0; i < airports.length(); i++) {
            Airport airport = new Airport(airports.getJSONObject(i));
            TerminalControl.radarScreen.airports.put(airport.getIcao(), airport);
            airport.loadOthers(airports.getJSONObject(i));
        }
    }
}

File: core/src/com/bombbird/terminalcontrol/utilities/files/GameLoader.kt
package com.bombbird.terminalcontrol.utilities.files

import com.badlogic.gdx.Gdx
import com.bombbird.terminalcontrol.TerminalControl
import com.bombbird.terminalcontrol.entities.aircrafts.Arrival
import com.bombbird.terminalcontrol.entities.aircrafts.Departure
import com.bombbird.terminalcontrol.entities.airports.Airport
import com.bombbird.terminalcontrol.entities.weather.ThunderCell
import com.bombbird.terminalcontrol.screens.gamescreen.RadarScreen
import com.bombbird.terminalcontrol.utilities.RenameManager.renameAirportICAO
import org.json.JSONArray
import org.json.JSONObject
import java.util.*

object GameLoader {
    /** Loads save information from the save JSONObject  */
    fun loadSaveData(save: JSONObject?) {
        val radarScreen = TerminalControl.radarScreen!!
        if (save == null) return

        val airports = save.getJSONArray("airports")
        for (airport in radarScreen.airports.values) {
            for (runway in airport.runways.values) {
                runway.label.remove()
            }
        }

        val jsonArray = save.getJSONArray("allAircraft")
        val allAircrafts = HashSet<String>()
        for (i in 0 until jsonArray.length()) {
            allAircrafts.add(jsonArray.getString(i))
        }
        radarScreen.allAircraft = allAircrafts

        loadAirportData(airports)
        radarScreen.metar.updateMetar(false)
        loadAircraft(save.getJSONArray("aircrafts"))
        for (i in 0 until airports.length()) {
            val airport = radarScreen.airports[renameAirportICAO(airports.getJSONObject(i).getString("icao"))]
            airport?.takeoffManager?.updatePrevAcft(airports.getJSONObject(i).getJSONObject("takeoffManager"))
            airport?.updateOtherRunwayInfo(airports.getJSONObject(i))
        }
        radarScreen.utilityBox.loadSave(save.getJSONArray("commBox"))
        radarScreen.separationChecker.lastNumber = save.getInt("lastNumber")
        radarScreen.separationChecker.time = save.optDouble("sepTime", 3.0).toFloat()

        radarScreen.playTime = save.optDouble("playTime", estimatePlayTime(radarScreen).toDouble()).toFloat()

        radarScreen.handoverController.loadSaveData(save.optJSONObject("handoverController"))

        val array = save.optJSONArray("thunderCells")
        if (array != null) {
            for (i in 0 until array.length()) {
                radarScreen.thunderCellArray.add(ThunderCell(array.optJSONObject(i)))
            }
        }

        //GameSaver.saveGame();
    }

    /** Estimates the duration played (if save has no play time data)  */
    private fun estimatePlayTime(radarScreen: RadarScreen): Float {
        var landed = 0
        for (airport in radarScreen.airports.values) {
            landed += airport.landings
        }

        //Assume 90 seconds between landings lol
        return (landed * 90).toFloat()
    }

    /** Loads aircraft data from save  */
    private fun loadAircraft(aircrafts: JSONArray) {
        val radarScreen = TerminalControl.radarScreen!!
        for (aircraft in radarScreen.aircrafts.values) {
            aircraft.removeAircraft()
        }
        for (i in 0 until aircrafts.length()) {
            when (aircrafts.getJSONObject(i).getString("TYPE")) {
                "Arrival" -> {
                    //Load arrival
                    val arrival = Arrival(aircrafts.getJSONObject(i))
                    radarScreen.aircrafts[arrival.callsign] = arrival
                }
                "Departure" -> {
                    //Load departure
                    val departure = Departure(aircrafts.getJSONObject(i))
                    radarScreen.aircrafts[departure.callsign] = departure
                }
                //Unknown type TODO En-route aircraft in future
                else -> Gdx.app.log("Aircraft load error", "Unknown aircraft type " + aircrafts.getJSONObject(i).getString("TYPE") + " in save file!")
            }
        }
    }

    /** Loads airport data from save  */
    private fun loadAirportData(airports: JSONArray) {
        for (i in 0 until airports.length()) {
            val airport = Airport(airports.getJSONObject(i))
            TerminalControl.radarScreen?.airports?.set(airport.icao, airport)
            airport.loadOthers(airports.getJSONObject(i))
        }
    }
}